File: research_agent.py
import os
import json
import time
from datetime import datetime
from typing import Dict, List, Optional, Callable
import requests
from web_scraper import WebScraper
import google.generativeai as genai

class ResearchAgent:
    """AI Research Agent for comprehensive company analysis"""
    
    def __init__(self, gemini_api_key: str, tavily_api_key: str):
        self.gemini_api_key = gemini_api_key
        self.tavily_api_key = tavily_api_key
        self.web_scraper = WebScraper()
        
        # Configure Gemini AI
        genai.configure(api_key=gemini_api_key)
        self.model = genai.GenerativeModel('gemini-1.5-flash')
        
    def research_company(
        self, 
        website_url: str, 
        max_pages: int = 5,
        include_external: bool = True,
        progress_callback: Optional[Callable] = None
    ) -> Dict:
        """
        Perform comprehensive company research
        
        Args:
            website_url: Target company website URL
            max_pages: Maximum number of pages to scrape
            include_external: Whether to include external research
            progress_callback: Function to call with progress updates
            
        Returns:
            Structured company profile dictionary
        """
        
        def update_progress(message: str, percentage: int):
            if progress_callback:
                progress_callback(message, percentage)
        
        try:
            # Step 1: Scrape website content
            update_progress("🔍 Scraping website content...", 30)
            scraped_data = self.web_scraper.scrape_website(website_url, max_pages)
            
            # Step 2: External research (if enabled)
            external_data = {}
            if include_external:
                update_progress("🌐 Conducting external research...", 50)
                external_data = self._perform_external_research(website_url, scraped_data.get('company_name', ''))
            
            # Step 3: Synthesize with AI
            update_progress("🤖 Synthesizing data with AI...", 70)
            company_profile = self._synthesize_with_ai(scraped_data, external_data)
            
            # Step 4: Finalize and structure
            update_progress("📊 Finalizing report...", 90)
            final_profile = self._finalize_profile(company_profile, website_url)
            
            update_progress("✅ Research completed!", 100)
            return final_profile
            
        except Exception as e:
            raise Exception(f"Research failed: {str(e)}")
    
    def _perform_external_research(self, website_url: str, company_name: str) -> Dict:
        """Perform external research using Tavily API"""
        
        if not company_name:
            # Extract company name from URL
            from urllib.parse import urlparse
            domain = urlparse(website_url).netloc
            company_name = domain.replace('www.', '').split('.')[0]
        
        try:
            # Tavily API search
            headers = {
                'Content-Type': 'application/json',
            }
            
            # Search queries for comprehensive research
            queries = [
                f"{company_name} company funding valuation",
                f"{company_name} founders executives leadership",
                f"{company_name} products services business model",
                f"{company_name} headquarters location industry"
            ]
            
            external_results = {
                'funding_info': [],
                'leadership_info': [],
                'business_info': [],
                'company_info': []
            }
            
            for i, query in enumerate(queries):
                try:
                    payload = {
                        'api_key': self.tavily_api_key,
                        'query': query,
                        'search_depth': 'basic',
                        'include_answer': True,
                        'include_raw_content': False,
                        'max_results': 3
                    }
                    
                    response = requests.post(
                        'https://api.tavily.com/search',
                        headers=headers,
                        json=payload,
                        timeout=10
                    )
                    
                    if response.status_code == 200:
                        data = response.json()
                        result_key = list(external_results.keys())[i]
                        external_results[result_key] = data.get('results', [])
                    
                    time.sleep(0.5)  # Rate limiting
                    
                except Exception as e:
                    print(f"Tavily search failed for query '{query}': {str(e)}")
                    continue
            
            return external_results
            
        except Exception as e:
            print(f"External research failed: {str(e)}")
            return {}
    
    def _synthesize_with_ai(self, scraped_data: Dict, external_data: Dict) -> Dict:
        """Use Gemini AI to synthesize all collected data"""
        
        # Prepare the prompt
        prompt = self._create_synthesis_prompt(scraped_data, external_data)
        
        try:
            # Generate response with Gemini
            response = self.model.generate_content(prompt)
            
            # Parse JSON response
            response_text = response.text.strip()
            
            # Clean up response if it contains markdown code blocks
            if response_text.startswith('```json'):
                response_text = response_text[7:]
            if response_text.endswith('```'):
                response_text = response_text[:-3]
            
            # Parse JSON
            company_profile = json.loads(response_text)
            return company_profile
            
        except json.JSONDecodeError as e:
            raise Exception(f"Failed to parse AI response as JSON: {str(e)}")
        except Exception as e:
            raise Exception(f"AI synthesis failed: {str(e)}")
    
    def _create_synthesis_prompt(self, scraped_data: Dict, external_data: Dict) -> str:
        """Create a comprehensive prompt for AI synthesis"""
        
        prompt = f"""
You are an expert business analyst. Analyze the following company data and create a comprehensive company profile in JSON format.

SCRAPED WEBSITE DATA:
{json.dumps(scraped_data, indent=2)}

EXTERNAL RESEARCH DATA:
{json.dumps(external_data, indent=2)}

Based on this information, create a structured company profile with the following JSON schema:

{{
  "company_name": "string",
  "website_url": "string", 
  "linkedin_url": "string or null",
  "confidence_score": "High/Medium/Low",
  "summary": {{
    "about": "comprehensive company description",
    "tagline": "company tagline or mission statement"
  }},
  "company_details": {{
    "industry": "primary industry",
    "founded_year": "number or null",
    "company_type": "Public/Private/Startup/etc",
    "headquarters": "city, country/state"
  }},
  "people": {{
    "founders": ["list of founder names"],
    "key_executives": ["list of key executives with titles"]
  }},
  "offerings": {{
    "service_details": ["list of main products/services"],
    "pricing_model": "description of pricing approach"
  }},
  "valuation_and_revenue": {{
    "value": "string representation of value",
    "metric_type": "valuation/revenue/funding",
    "source": "source of information",
    "source_url": "url of source or null",
    "date_of_metric": "YYYY-MM-DD or null",
    "explanation": "brief explanation of the metric"
  }},
  "contact_info": {{
    "phone": "phone number or null",
    "email": "email address or null", 
    "contact_page_url": "contact page URL or null"
  }},
  "reference_links": {{
    "crunchbase_url": "crunchbase profile URL or null",
    "wikipedia_url": "wikipedia page URL or null",
    "other": ["array of other relevant URLs"]
  }},
  "last_updated": "YYYY-MM-DD"
}}

INSTRUCTIONS:
1. Extract and synthesize information from both scraped and external data
2. If information is missing or unclear, use null values
3. Provide a confidence score based on data quality and completeness
4. Ensure all extracted information is factual and verifiable
5. Use today's date for last_updated field
6. Return ONLY valid JSON, no additional text or explanations

JSON Response:
"""
        
        return prompt
    
    def _finalize_profile(self, profile: Dict, website_url: str) -> Dict:
        """Finalize the company profile with additional processing"""
        
        # Ensure required fields
        profile['website_url'] = website_url
        profile['last_updated'] = datetime.now().strftime('%Y-%m-%d')
        
        # Set confidence score if not provided
        if 'confidence_score' not in profile:
            profile['confidence_score'] = 'Medium'
        
        # Ensure nested dictionaries exist
        required_sections = [
            'summary', 'company_details', 'people', 
            'offerings', 'contact_info', 'reference_links'
        ]
        
        for section in required_sections:
            if section not in profile:
                profile[section] = {}
        
        return profile

File: netlify/functions/research-agent.js
const https = require('https');
const http = require('http');
const { URL } = require('url');

// Helper function to make HTTP requests
function makeRequest(url, options = {}) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);
        const lib = urlObj.protocol === 'https:' ? https : http;
        
        const requestOptions = {
            hostname: urlObj.hostname,
            port: urlObj.port,
            path: urlObj.pathname + urlObj.search,
            method: options.method || 'GET',
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
                'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
                ...options.headers
            }
        };

        const req = lib.request(requestOptions, (res) => {
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                resolve({
                    statusCode: res.statusCode,
                    headers: res.headers,
                    body: data
                });
            });
        });

        req.on('error', (err) => {
            reject(err);
        });

        if (options.body) {
            req.write(options.body);
        }

        req.end();
    });
}

// Extract text content from HTML
function extractTextFromHTML(html) {
    // Simple HTML tag removal - basic implementation
    let text = html.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '');
    text = text.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '');
    text = text.replace(/<[^>]*>/g, ' ');
    text = text.replace(/\s+/g, ' ').trim();
    return text;
}

// Extract company name from URL and content
function extractCompanyName(url, content) {
    try {
        const urlObj = new URL(url);
        const domain = urlObj.hostname.replace('www.', '');
        const domainName = domain.split('.')[0];
        
        // Look for title tags or h1 elements in content
        const titleMatch = content.match(/<title[^>]*>([^<]+)<\/title>/i);
        if (titleMatch) {
            const title = titleMatch[1].trim();
            if (title.length > 0 && title.length < 100) {
                return title.split(' - ')[0].split(' | ')[0];
            }
        }
        
        return domainName.charAt(0).toUpperCase() + domainName.slice(1);
    } catch (e) {
        return 'Unknown Company';
    }
}

// Extract emails from content
function extractEmails(content) {
    const emailRegex = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g;
    const emails = content.match(emailRegex) || [];
    
    // Filter out common non-business emails
    const excludedDomains = ['example.com', 'test.com', 'gmail.com', 'yahoo.com'];
    return emails.filter(email => {
        const domain = email.split('@')[1].toLowerCase();
        return !excludedDomains.some(excluded => domain.includes(excluded));
    });
}

// Perform external research using Tavily API
async function performExternalResearch(companyName, tavilyApiKey) {
    if (!tavilyApiKey) {
        return { error: 'Tavily API key not provided' };
    }

    try {
        const queries = [
            `${companyName} company funding valuation`,
            `${companyName} founders executives leadership`,
            `${companyName} products services business model`
        ];

        const results = {
            funding_info: [],
            leadership_info: [],
            business_info: []
        };

        for (let i = 0; i < queries.length; i++) {
            try {
                const response = await makeRequest('https://api.tavily.com/search', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify({
                        api_key: tavilyApiKey,
                        query: queries[i],
                        search_depth: 'basic',
                        include_answer: true,
                        include_raw_content: false,
                        max_results: 3
                    })
                });

                if (response.statusCode === 200) {
                    const data = JSON.parse(response.body);
                    const resultKey = Object.keys(results)[i];
                    results[resultKey] = data.results || [];
                }
            } catch (e) {
                console.log(`Tavily search failed for query: ${queries[i]}`);
            }
        }

        return results;
    } catch (error) {
        return { error: error.message };
    }
}

// Synthesize data with Gemini AI
async function synthesizeWithAI(scrapedData, externalData, geminiApiKey) {
    if (!geminiApiKey) {
        throw new Error('Gemini API key not provided');
    }

    const prompt = `
You are an expert business analyst. Analyze the following company data and create a comprehensive company profile in JSON format.

SCRAPED WEBSITE DATA:
${JSON.stringify(scrapedData, null, 2)}

EXTERNAL RESEARCH DATA:
${JSON.stringify(externalData, null, 2)}

Based on this information, create a structured company profile with the following JSON schema:

{
  "company_name": "string",
  "website_url": "string", 
  "linkedin_url": "string or null",
  "confidence_score": "High/Medium/Low",
  "summary": {
    "about": "comprehensive company description",
    "tagline": "company tagline or mission statement"
  },
  "company_details": {
    "industry": "primary industry",
    "founded_year": "number or null",
    "company_type": "Public/Private/Startup/etc",
    "headquarters": "city, country/state"
  },
  "people": {
    "founders": ["list of founder names"],
    "key_executives": ["list of key executives with titles"]
  },
  "offerings": {
    "service_details": ["list of main products/services"],
    "pricing_model": "description of pricing approach"
  },
  "valuation_and_revenue": {
    "value": "string representation of value",
    "metric_type": "valuation/revenue/funding",
    "source": "source of information",
    "source_url": "url of source or null",
    "date_of_metric": "YYYY-MM-DD or null",
    "explanation": "brief explanation of the metric"
  },
  "contact_info": {
    "phone": "phone number or null",
    "email": "email address or null", 
    "contact_page_url": "contact page URL or null"
  },
  "reference_links": {
    "crunchbase_url": "crunchbase profile URL or null",
    "wikipedia_url": "wikipedia page URL or null",
    "other": ["array of other relevant URLs"]
  },
  "last_updated": "YYYY-MM-DD"
}

INSTRUCTIONS:
1. Extract and synthesize information from both scraped and external data
2. If information is missing or unclear, use null values
3. Provide a confidence score based on data quality and completeness
4. Ensure all extracted information is factual and verifiable
5. Use today's date for last_updated field
6. Return ONLY valid JSON, no additional text or explanations

JSON Response:`;

    try {
        const response = await makeRequest(`https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash-latest:generateContent?key=${geminiApiKey}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                contents: [{
                    parts: [{
                        text: prompt
                    }]
                }],
                generationConfig: {
                    temperature: 0.1,
                    topK: 1,
                    topP: 1,
                    maxOutputTokens: 2048,
                }
            })
        });

        if (response.statusCode !== 200) {
            throw new Error(`Gemini API error: ${response.statusCode}`);
        }

        const data = JSON.parse(response.body);
        let responseText = data.candidates[0].content.parts[0].text;

        // Clean up response if it contains markdown code blocks
        if (responseText.startsWith('```json')) {
            responseText = responseText.slice(7);
        }
        if (responseText.endsWith('```')) {
            responseText = responseText.slice(0, -3);
        }

        const companyProfile = JSON.parse(responseText);
        return companyProfile;

    } catch (error) {
        throw new Error(`AI synthesis failed: ${error.message}`);
    }
}

// Main handler function
exports.handler = async (event, context) => {
    // Enable CORS
    const headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS'
    };

    // Handle preflight requests
    if (event.httpMethod === 'OPTIONS') {
        return {
            statusCode: 200,
            headers,
            body: ''
        };
    }

    if (event.httpMethod !== 'POST') {
        return {
            statusCode: 405,
            headers,
            body: JSON.stringify({ error: 'Method not allowed' })
        };
    }

    try {
        const { websiteUrl, maxPages = 5, includeExternal = true } = JSON.parse(event.body);

        if (!websiteUrl) {
            return {
                statusCode: 400,
                headers,
                body: JSON.stringify({ error: 'Website URL is required' })
            };
        }

        // Get environment variables
        const geminiApiKey = process.env.GEMINI_API_KEY;
        const tavilyApiKey = process.env.TAVILY_API_KEY;

        if (!geminiApiKey) {
            return {
                statusCode: 400,
                headers,
                body: JSON.stringify({ error: 'Gemini API key not configured' })
            };
        }

        // Step 1: Scrape website content
        const response = await makeRequest(websiteUrl);
        if (response.statusCode >= 400) {
            throw new Error(`Failed to fetch website: ${response.statusCode}`);
        }

        const htmlContent = response.body;
        const textContent = extractTextFromHTML(htmlContent);
        
        const scrapedData = {
            base_url: websiteUrl,
            company_name: extractCompanyName(websiteUrl, htmlContent),
            content: textContent.substring(0, 5000), // Limit content length
            emails: extractEmails(textContent),
            scraped_pages_count: 1
        };

        // Step 2: External research (if enabled)
        let externalData = {};
        if (includeExternal && tavilyApiKey) {
            externalData = await performExternalResearch(scrapedData.company_name, tavilyApiKey);
        }

        // Step 3: Synthesize with AI
        const companyProfile = await synthesizeWithAI(scrapedData, externalData, geminiApiKey);

        // Step 4: Finalize profile
        companyProfile.website_url = websiteUrl;
        companyProfile.last_updated = new Date().toISOString().split('T')[0];

        if (!companyProfile.confidence_score) {
            companyProfile.confidence_score = 'Medium';
        }

        return {
            statusCode: 200,
            headers,
            body: JSON.stringify(companyProfile)
        };

    } catch (error) {
        console.error('Research error:', error);
        return {
            statusCode: 500,
            headers,
            body: JSON.stringify({ 
                error: error.message || 'An unexpected error occurred during research'
            })
        };
    }
};